### PR TITLE
implement new sniff to force spaces after `=`

### PR DESCRIPTION
**Pull request by Ocramius**
I haven't yet found the rule that will help with that (since `MultipleStatementAlignmentSniff` doesn't work here)

This PR/issue comes from this #103 (comment)

This is a copy of https://github.com/doctrine/coding-standard/pull/117 as per https://github.com/doctrine/coding-standard/pull/117#issuecomment-488038466

### CODE REVIEW

**Comment by Ocramius** [comment]
Travis-CI really doesn't like this patch at all: I'll push it up as a branch of this repo.

**Comment by SenseException** [comment]
"Some checks haven’t completed yet" :-/

**Comment by SenseException** [comment]
If this is going to be the first custom code for the coding-standard, should it introduce the test environment like introduced in #63?

File: lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $pointer) : void
+    {
+        if (! $this->isOperator($phpcsFile, $pointer)) {
+            return;
+        }
+
+        $tokens = $phpcsFile->getTokens();
```
Comment (greg0ire) [inline lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php]:
```suggestion
```

File: lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        $this->ensureOneSpaceAfterOperator($phpcsFile, $pointer, $tokens);
```
Comment (greg0ire) [inline lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php]:
```suggestion
        $this->ensureOneSpaceAfterOperator($phpcsFile, $pointer, $phpcsFile->getTokens());
```

File: lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php (hunk 1 above)
Comment (vladyslavstartsev) [reply lib/Doctrine/Sniffs/Arrays/OperatorSpacingAfterSniff.php]: thanks! fixed.

**Comment by vladyslavstartsev** [comment]
> If this is going to be the first custom code for the coding-standard, should it introduce the test environment like introduced in #63?

I can add if you want

**Comment by Ocramius** [comment]
Yes, we'd most likely need proper coverage for new code, highlighting how much we're covering (branch-related). The existing tests are for regression/integration, and don't give us this confidence.

**Comment by vladyslavstartsev** [comment]
@malarzm yes, you are absolutely right! Thanks for mentioning this

File: tests/input/trailing_comma_on_array.php [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 declare(strict_types=1);
 
-$array =  [
```
Comment (vladyslavstartsev) [inline tests/input/trailing_comma_on_array.php]: I've found the error in other test file, so I've fixed it.

File: composer.json [hunk 4]
```diff
@@ -38,5 +38,14 @@
         "branch-alias": {
             "dev-master": "7.0.x-dev"
         }
+    },
+    "require-dev": {
+        "phpunit/phpunit": "^7.2",
+        "phpstan/phpstan": "^0.10"
```
Comment (Ocramius) [inline composer.json]: We should pin to newer releases

File: phpstan.neon.dist [hunk 5]
```diff
@@ -0,0 +1,6 @@
+parameters:
+        level: 7
+        paths:
+        - "%currentWorkingDirectory%/lib/"
```
Comment (Ocramius) [inline phpstan.neon.dist]: Also need to cover tests: phpstan-phpunit may be needed

File: tests/bootstrap.php [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?php
```
Comment (Ocramius) [inline tests/bootstrap.php]: Is this file actually needed, considering composer's autoloader?

File: tests/Doctrine/Sniffs/Arrays/data/operatorSpacingAfterSniff.errors.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+
+$arrayWithMultipleSpacesAfterAssigmentOperator =                       [
+    0,
+    1,
+    2,
+    3,
+    4,
+    5,
+];
```
Comment (Ocramius) [inline tests/Doctrine/Sniffs/Arrays/data/operatorSpacingAfterSniff.errors.php]: We should probably keep all other CS rules consistent in these files, to highlight the actual issue. 

In this case, we miss strict type declaration and the EOL

File: tests/bootstrap.php (hunk 6 above)
Comment (vladyslavstartsev) [reply tests/bootstrap.php]: yes it is, see https://github.com/slevomat/coding-standard/blob/1ff7cfb8e8df3af138cd04793ca1fbfea254cd50/phpunit.xml.dist#L3

Code_sniffer has custom auto loading on tests

File: tests/Doctrine/Sniffs/Arrays/data/operatorSpacingAfterSniff.errors.php (hunk 7 above)
Comment (vladyslavstartsev) [reply tests/Doctrine/Sniffs/Arrays/data/operatorSpacingAfterSniff.errors.php]: Yes, I did run `phpcs` over it, but somewhy it seems not working for me. Any idea?

```bash
root@f464869d226f:/var/www$ ./vendor/bin/phpcs tests/Doctrine/Sniffs/Arrays/data/ 


Time: 52ms; Memory: 8MB

```

Comment (vladyslavstartsev) [reply tests/Doctrine/Sniffs/Arrays/data/operatorSpacingAfterSniff.errors.php]: anyway, I'll fix the EOL and strict types issue

**Comment by vladyslavstartsev** [comment]
@Ocramius ok, I've done all the things, but one thing that doesn't work as expected is  https://github.com/doctrine/coding-standard/pull/123#discussion_r283136248

File: phpstan.neon.dist (hunk 5 above)
Comment (vladyslavstartsev) [reply phpstan.neon.dist]: added. Done

File: composer.json (hunk 4 above)
Comment (vladyslavstartsev) [reply composer.json]: done

**Comment by vladyslavstartsev** [comment]
> Apart from missing newlines in some files it's  from me

:man_facepalming: yes, I see, I can fix it, but it will dismiss all of the reviews (don't want bother everyone yet another time)

**Comment by deeky666** [comment]
> > Apart from missing newlines in some files it's  from me
> 
> man_facepalming yes, I see, I can fix it, but it will dismiss all of the reviews (don't want bother everyone yet another time)

that's the way it is, no worries :)

File: composer.json [hunk 8]
```diff
@@ -38,5 +38,17 @@
         "branch-alias": {
             "dev-master": "7.0.x-dev"
         }
+    },
+    "require-dev": {
+        "phpunit/phpunit": "^8.1",
```
Comment (carusogabriel) [inline composer.json]: As we minimum required PHP 7.1, we can't use PHPUnit 8. Can you downgrade to PHPUnit 7, please?

Comment (vladyslavstartsev) [reply composer.json]: Sure :man_facepalming: my bad

Comment (vladyslavstartsev) [reply composer.json]: did that

**Comment by carusogabriel** [comment]
@vladyslavstartsev Please, see the [tests failures](https://travis-ci.org/doctrine/coding-standard/jobs/533444151).

**Comment by vladyslavstartsev** [comment]
Can you retrigger Travis CI?

File: docs/en/reference/index.rst [hunk 9]
```diff
@@ -132,3 +132,15 @@ need to execute PHP_CodeSniffer with the tests folder and ensure it matches the
 .. code-block:: bash
 
     $ ./vendor/bin/phpcs tests/input --report=summary --report-file=phpcs.log; diff tests/expected_report.txt phpcs.log
+
+If you have implemented custom sniff also run PHPStan and and PHPUnit.
```
Comment (carusogabriel) [inline docs/en/reference/index.rst]:
```suggestion
If you have implemented custom sniff also run PHPStan and PHPUnit.
```

File: phpstan.neon.dist [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+        ignoreErrors:
+                -
+                        message: '#Method [a-zA-Z0-9\\_]+::process\(\) has parameter \$pointer with no typehint specified#'
+                        path: %currentWorkingDirectory%/lib/Doctrine/Sniffs/*
+
+includes:
+        - vendor/phpstan/phpstan-deprecation-rules/rules.neon
+        - vendor/phpstan/phpstan-phpunit/extension.neon
+        - vendor/phpstan/phpstan-phpunit/rules.neon
+        - vendor/phpstan/phpstan-strict-rules/rules.neon
```
Comment (deeky666) [inline phpstan.neon.dist]: Still missing a newline

**Comment by malarzm** [comment]
`=&` looks like an operator to me, personally I prefer writing `$this->foo[$bar] = &$baz['quox'];`

**Comment by carusogabriel** [comment]
My bad, I was confusing this with `+=`, `*=`, `.=` :man_facepalming:

**Comment by vladyslavstartsev** [comment]
@carusogabriel since there is #126 should I close this one? Or make any changes?

**Comment by carusogabriel** [comment]
@vladyslavstartsev Let's hold on this one until we get slevomat/cs v6 released, as well PHPCS v3.5 :)

**Comment by vladyslavstartsev** [comment]
@carusogabriel sure! :+1:

**Comment by simPod** [comment]
We have created similar sniff a while ago

See  
https://github.com/cdn77/coding-standard/blob/master/tests/Sniffs/Operators/data/OperatorSpacingSniffErrors.php
https://github.com/cdn77/coding-standard/blob/master/tests/Sniffs/Operators/data/OperatorSpacingSniffErrors.fixed.php

I wonder whether it would be possible to maybe reuse it here in some way?

**Comment by carusogabriel** [comment]
@simPod Looks like an idea. I hoped that PHPCS v3.5 would bring a fix to #126, but it looks like that didn't happen. 

Unfortunately, we can't require `cdn77/coding-standard` here as that depends on `doctrine/coding-standard`, and it will always conflict with the `master` version.

Fell free to open a PR copying and pasting the sniff you created here if you allow us to use it, while we don't have the sniff of #126 ready to use 👍

**Comment by grongor** [comment]
Hi, I created the PR @carusogabriel requested, based on @simPod suggestion. Check it out here https://github.com/doctrine/coding-standard/pull/134 :)
